[FLINK-25431] Implement a file-based JobResultStore

autophagy opened the pull request:
This commit contains an implementation of the JobResultStore backed by a
file-system. By default, this filesystem is based on the settings for
high-availability.storageDir and the cluster-id, but can be overriden
with the job-result-store.storage-path setting.

Alongside this setting, it also introduces a new option
job-result-store.delete-on-commit, which determines whether the
underlying file for the job result entry is deleted when it is clean, or
whether it is renamed and persisted. By default, this option is true.

Here is a CI run for this PR: https://dev.azure.com/Autophagy/Flink-CI/_build/results?buildId=191&view=results

XComp committed:
[hotfix] Makes intention of comment clearer

XComp committed:
[hotfix] Adds missing JavaDoc

XComp committed:
[FLINK-25430] Renames JobGraphStoreFactory into JobPersistenceComponentFactory

XComp committed:
[FLINK-25430] Make JobResultStore available to SessionDispatcherLeaderProcess

A new method HaServicesJobPersistenceComponentFactory.createJobResultStore was
introduced to achieve that. The goal is to initialize the JobResultStore
analogously to how the JobGraphWriter is initialized. Both resources are used
to recovere job data after a failover which makes them semantically similar.

XComp committed:
[FLINK-25430] Renames PartialDispatcherServicesWithJobGraphStore into PartialDispatcherServicesWithJobPersistenceComponents

XComp committed:
[hotfix] Removes @Nonnull annotations from DispatcherServices and PartialDispatcherServicesWithJobGraphStore

Instead, null checks are added

XComp committed:
[hotfix] Adds fail to DispatcherTest

XComp committed:
[hotfix] Migrates ApplicationStatusTest to JUnit5 and AssertJ

autophagy committed:
[FLINK-25431] Implement a file-based JobResultStore

This commit contains an implementation of the JobResultStore backed by a
file-system. By default, this filesystem is based on the settings for
high-availability.storageDir and the cluster-id, but can be overriden
with the job-result-store.storage-path setting.

Alongside this setting, it also introduces a new option
job-result-store.delete-on-commit, which determines whether the
underlying file for the job result entry is deleted when it is clean, or
whether it is renamed and persisted. By default, this option is true.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/JobResultEntry.java:
```diff
@@ -18,12 +18,18 @@
  */
+@JsonSerialize(using = JobResultEntrySerializer.class)
+@JsonDeserialize(using = JobResultEntryDeserializer.class)
 public class JobResultEntry implements WithJobResult {
```
Have you had a look at [JobExecutionResultResponseBody](https://github.com/apache/flink/blob/c6997c97c575d334679915c328792b8a3067cfb5/flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/JobExecutionResultResponseBody.java)? You could define the serialization/deserialization entirely using annotations like `@JsonProperty`. There's no need to come up with your own Serializer/Deserializer implementation as far as I can see. Or did you consider that already and I'm just missing something?

autophagy replied:
I wanted to do something like this, but I assumed that the requirement that the JSON payload be versioned and that the deserializer be chosen based on that version value would require a custom serializer? I might be mistaken, I'm mostly working from prior art in Flink in how to do this :sweat_smile:

XComp replied:
I was thinking about it once more. You could create a subclass of `JobResultEntry` called something `JsonJobResultStoreEntry` that provides a copy constructor and just adds an additional member for the serialization version. This one you could annotate with the Jackson annotations in the end and avoid using the custom serialization. WDYT? 🤔

autophagy replied:
I think that would be nicer! I'm still unsure how we would then go about selecting different deserialization strategies based on that version, though. I'll have a look through the codebase to see if something similar is being used elsewhere.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,203 @@
+
+    private static final String DIRTY_SUFFIX = "_DIRTY";
+
+    public FileSystemJobResultStore(Configuration config) throws IOException {
```
Maybe, that's a personal opinion, but: It's always nice to have a default constructor with the actually required parameters (in this case, `basePath`, `deleteOnCommit` and probably even `FileSystem`) instead of only having a constructor that generates the member variables from some other object. That constructor can have limited visibility for testing purposes. It improves testability of the code (e.g. using a dummy FileSystem instead of the actual file system; I didn't check, though, whether that makes sense in this specific case). The configuration extraction can be done in a separate method which can be tested individually.

...just as a general idea. Here, it's not that big of an issue because instances of type `Configuration` can be easily created. But I still think that it's a good practise to consider when creating new classes.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,203 @@
+        Path dirtyPath = constructDirtyPath(jobId);
+        if (fileSystem.exists(dirtyPath)) {
+            if (deleteOnCommit) {
+                fileSystem.delete(constructDirtyPath(jobId), false);
```
here, we recreate the dirty path even though we have it already stored in `dirtyPath`

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,203 @@
+            } else {
+                JobResultEntry resultEntry =
+                        mapper.readValue(new File(dirtyPath.getPath()), JobResultEntry.class);
+                OutputStream os = fileSystem.create(dirtyPath, FileSystem.WriteMode.OVERWRITE);
```
What's the purpose of reading and writing the entry here? The `JobResultEntry` didn't change... 🤔

autophagy replied:
Ah, I think this is a holdover from before we removed the dirty/clean state being part of the job result entry. Before, it'd have to read, modify the state, and write it back. Thanks for pointing this out, I completely missed it.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,203 @@
+                        mapper.readValue(new File(dirtyPath.getPath()), JobResultEntry.class);
+                OutputStream os = fileSystem.create(dirtyPath, FileSystem.WriteMode.OVERWRITE);
+                mapper.writeValue(os, resultEntry);
+                fileSystem.rename(constructDirtyPath(jobId), constructCleanPath(jobId));
```
There is no indication in the `FileSystem`'s JavaDoc that the rename operation is thread-safe. So, we would need to ensure thread-safety here.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,203 @@
+ * An implementation of the {@link JobResultStore} which persists job result data to an underlying
+ * distributed filesystem.
+ */
+public class FileSystemJobResultStore implements JobResultStore {
```
`FileSystemJobResultStore` is not thread-safe. Each operation should be done in an atomic manner. The goal is to not end up in an inconsistent state while marking a job as clean. You might have a look at the [FLINK-25430 PR](https://github.com/apache/flink/pull/18189) on how the thread-safety is done in the `EmbeddedJobResultStore`.

XComp replied:
Looks good/ One other thing we could have done is to move the synchronization logic (and probably also some of the state checks) into an `abstract` parent class.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/JobResultEntryDeserializer.java:
```diff
@@ -0,0 +1,85 @@
+import static org.apache.flink.runtime.highavailability.JobResultEntrySerializer.FIELD_NAME_VERSION;
+
+/** JSON serializer for {@link JobResultEntry}. */
+public class JobResultEntryDeserializer extends StdDeserializer<JobResultEntry> {
```
Same as what I already wrote for the `JobResultEntry` class changes itself: We might want to use the default deserialization functionality of Jackson. ...if I didn't miss any other reason why it's necessary.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/JobResultEntrySerializer.java:
```diff
@@ -0,0 +1,67 @@
+import java.io.IOException;
+
+/** JSON serializer for {@link JobResultEntry}. */
+public class JobResultEntrySerializer extends StdSerializer<JobResultEntry> {
```
Same as what I already wrote for the `JobResultEntry` class changes itself: We might want to use the default serialization functionality of Jackson. ...if I didn't miss any other reason why it's necessary.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/JobResultStoreOptions.java:
```diff
@@ -0,0 +1,48 @@
+    public static final ConfigOption<String> STORAGE_PATH =
+            key("job-result-store.storage-path")
+                    .stringType()
+                    .defaultValue(null)
```
Doesn't that fail (I didn't check the test run but checked [ConfigOptions:181](https://github.com/apache/flink/blob/c6997c97c575d334679915c328792b8a3067cfb5/flink-core/src/main/java/org/apache/flink/configuration/ConfigOptions.java#L181). You might want to use `noDefaultValue()` in that case.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/JobResultStoreOptions.java:
```diff
@@ -0,0 +1,48 @@
+public class JobResultStoreOptions {
+    @Documentation.Section(Documentation.Sections.COMMON_HIGH_AVAILABILITY)
+    public static final ConfigOption<String> STORAGE_PATH =
+            key("job-result-store.storage-path")
```
nit: In other `*Options` classes, they import `ConfigOptions` and initialize each option like `ConfigOption.key("...")...`.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/JobResultStoreOptions.java:
```diff
@@ -0,0 +1,48 @@
+
+/** The set of configuration options relating to the Job Result Store. */
+public class JobResultStoreOptions {
+    @Documentation.Section(Documentation.Sections.COMMON_HIGH_AVAILABILITY)
```
We have to regenerate the Flink docs to make this change endup in the docs (see [./docs/README.md:35](https://github.com/apache/flink/blob/dd48d058c6b745f505870836048284a76a23f7cc/docs/README.md#L35)). This will create extend the diff in this PR.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/JobResultStoreOptions.java:
```diff
@@ -0,0 +1,48 @@
+                    .withDescription(
+                            "Defines where job results should be stored. This should be an "
+                                    + "underlying file-system that provides read-after-write consistency. By "
+                                    + "default, uses a combination of high-availability.storageDir and "
+                                    + "high-availability.cluster-id to determine the path.");
```
Just an idea: What about moving the description (I'm not talking about the entire text here) of the default definition into the class where it's generated and make it available through a public static method? That way, we have the logic and the documentation in one place and reduce the risk of forgetting to update the documentation if we decide to change the implementation for the default generation.

XComp replied:
Have you thought about my remark?

autophagy replied:
I have! Would there be an objection to just moving the config options wholesale into the FilesystemJobResultStore class? If they are not used anywhere else, I feel like having a seperate class is not so necessary and it keeps the documentation closer to the implementation, as you mentioned.

autophagy replied:
@XComp Ah, it doesnt look like this works. I think when it generates the docs it looks in a specific package, and isnt picked up if its in the `FilesystemJobResultStore` class. What do you mean about "the description (not the entire text)"? I feel like having text in 2 places might be confusing - maybe it would be better to remove implementation details at all?

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/JobResultStoreOptions.java:
```diff
@@ -0,0 +1,48 @@
+                    .defaultValue(Boolean.TRUE)
+                    .withDescription(
+                            "Determines whether job results should be automatically removed "
+                                    + "from the filesystem when they are clean. If false, the cleaned job "
+                                    + "result files are instead renamed to indicate their cleaned state.");
```
nit: I'm wondering whether we want to mention the renaming here or whether it's an implementation detail. 🤔 A more general term `[...] are, instead, marked as clean to indicate their state.`

autophagy replied:
Yeah, I think so. Perhaps there shouldn't be any references to filesystem at all here, since that's technically and implementation detail and I think this setting would be used in any other JRS implementations that persist the data somewhere that isn't necessarily a filesystem.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/JobResultStoreOptions.java:
```diff
@@ -0,0 +1,48 @@
+                    .withDescription(
+                            "Determines whether job results should be automatically removed "
+                                    + "from the filesystem when they are clean. If false, the cleaned job "
+                                    + "result files are instead renamed to indicate their cleaned state.");
```
```suggestion
                                    + "from the filesystem when they are clean. If false, the cleaned job's "
                                    + "job result files are instead renamed to indicate their cleaned state.");
```
nit: Actually, the corresponding Job is cleaned up, not the file.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/JobResultEntry.java:
```diff
@@ -35,4 +41,20 @@ public JobResultEntry(JobResult jobResult) {
+
+    /** Builder for {@link JobResultEntry}. */
+    @Internal
+    public static class Builder {
```
Is the `Builder` necessary for the Json serialization? Otherwise, it's not adding much value since we don't have a default value for `JobResult` provided. Essentially, the Builder does the same like the constructor. WDYT?

Thinking about it, we might want to add a `NonNull` check to `JobResult` since it's more of a required value. I will add this in the downstream branch.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,185 @@
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
+
+/** Tests for the {@link FileSystemJobResultStore}. */
+public class FileSystemJobResultStoreTest extends TestLogger {
```
I remember parts of the code. :-D Initially, I was wondering whether we should combine `EmbeddedJobResultStoreTest` and `FileSystemJobResultStoreTest` to reduce code redundancy. But I'm not a big fan of coming up with abstract test classes because it makes reading the test class harder. Alternatively, we could refactor `EmbeddedJobResultStore` to provide static test method that can be called within `EmbeddedJobResultStoreTest` and `FileSystemJobResultStoreTest`. Not sure, maybe, @dmvk has a strong opinion on that one.

Anyway, the `FileSystemJobResultStoreTest` needs to check the actual file creation as well since it's part of the classes state.

autophagy replied:
Yeah, I think coming up with an abstract test class for only 2 test classes is maybe overkill. Could be something to look forward to if we implement another implementation down the line.

XComp replied:
Agree 👍

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,185 @@
+        Configuration config = new Configuration();
+        config.setString(
+                HighAvailabilityOptions.HA_STORAGE_PATH, temporaryFolder.newFolder().getPath());
+        config.setBoolean(JobResultStoreOptions.DELETE_ON_COMMIT, false);
```
We also want to check the `DELETE_ON_COMMIT` flag being `true`.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/JobResultEntryDeserializerTest.java:
```diff
@@ -0,0 +1,107 @@
+import static org.junit.Assert.assertThat;
+
+/** Tests for {@link JobResultEntryDeserializer}. */
+public class JobResultEntryDeserializerTest extends TestLogger {
```
We wouldn't need to test the exact serialization if we rely on Jackson's default serialization using annotations as described in previous comments. I'm wondering whether serializing an object and then deserializing its JSON representation would be enough. Essentially, we would want to have a copy of the object in the end. Something like this is done in `RestResponseMarshallingTestBase` for the REST API responses. WDYT? Or is there something else, I'm missing that requires us to look at the exact content of the Json? 🤔

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,203 @@
+            this.basePath = new Path(jrsStoragePath);
+        }
+
+        this.fileSystem = this.basePath.getFileSystem();
+        this.fileSystem.mkdirs(this.basePath);
```
It might be overkill, but: I was thinking about having wrapper around the FileSystem that takes care of creating and renaming files and checking for existence and dealing with the folder structure. `FileSystemJobResultStore` would be in charge of the serialization and utilizing this wrapper implementation. ...but that's just an idea I wanted to share. I haven't thought this through. ...especially considering that `FileSystemJobResultStore` is not too complex on its own. ¯\_(ツ)_/¯

XComp reviewed (changes requested):
Thanks @autophagy for the proposal. It looks promising. :-) I added a few comments below.

autophagy commented:
@XComp Thanks for the review! I think i've adressed most of your points.

The only main outstanding one is the serializer. I chose the manually write one over using annotations due to the requirement in the FLIP of a way of versioning the schemas/deserializers, to provide some backwards compatibility if the structure is ever changed in the future. What do you think about this? Do you know if theres a way of doing this only with annotations?

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,220 @@
+                HighAvailabilityOptions.HA_STORAGE_PATH, temporaryFolder.newFolder().getPath());
+        config.setBoolean(JobResultStoreOptions.DELETE_ON_COMMIT, false);
+
+        fileSystemJobResultStore = FileSystemJobResultStore.fromConfiguration(config);
```
nit: Here, we could use the constructor now making things more explicit.

autophagy replied:
Would it make sense to create a test for loading from configuration - given that the end result is setting some private fields, im not sure how straightforward that is?

XComp replied:
You have a good point. We might just leave it like that. Alternatively, you could add some package-private getter methods that you use to verify the initialized fields.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,220 @@
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
+
+/** Tests for the {@link FileSystemJobResultStore}. */
+public class FileSystemJobResultStoreTest extends TestLogger {
```
You might want to revisit the `JobResultStore` interface's JavaDoc to check for the exact contract. That changed slightly in the base branch (see [FLINK-25430 PR](https://github.com/apache/flink/pull/18189)). Feel free to revisit the `EmbeddedJobResultStoreTest` to identify test cases. They should be the same accept for the filesystem checks...

XComp replied:
```suggestion
@ExtendWith(TestLoggerExtension.class)
public class FileSystemJobResultStoreTest {
```
The JUnit5 way of doing things now (I missed that one as well :-D)

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,220 @@
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
```
We want to use Junit5 in new testcases

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,248 @@
+     * @return A path for a clean entry for the given the Job ID.
+     */
+    @VisibleForTesting
+    protected Path constructCleanPath(JobID jobId) {
```
Usually, it's enough to mark methods (or member variables) as package-protected (i.e. without any label) when making them visible for testing. In this specific case, you're using the method as some kind of helper method to construct the desired path and avoid creating redundant code. That's something you could do. 

But with the two `construct*Path` methods, you work around testing the actual path creation in `FileBasedJobResultStoreTest`. We're not testing the implementations of the two methods. Think about a case where one accidentally add the wrong file extension (`.txt` instead of `.json`) in the implementation of both `construct*Path`. That's not going to be reflected in the test.

I'd propose making these two methods `private` again and adding some helper method generating the expected path within `FileSystemJobResultStoreTest`.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,248 @@
+        if (isNullOrWhitespaceOnly(jrsStoragePath)) {
+            final String haStoragePath = config.get(HighAvailabilityOptions.HA_STORAGE_PATH);
+            final String haClusterId = config.get(HighAvailabilityOptions.HA_CLUSTER_ID);
+            basePath = new Path(haStoragePath + "/job-results-store/" + haClusterId);
```
nit: I know that that's the name being suggested in [FLIP-194](https://cwiki.apache.org/confluence/display/FLINK/FLIP-194%3A+Introduce+the+JobResultStore). But I'd suggest using `job-result-store` as the default folder name in order to align it with the interface's name

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,248 @@
+     * @param jobId The job ID to construct a dirty entry path from.
+     * @return A path for a dirty entry for the given the Job ID.
+     */
+    @VisibleForTesting
```
Analogously to the comment about the `constructCleanPath` below.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,248 @@
+
+    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+
+    private final FileSystem fileSystem;
```
```suggestion
    @GuardedBy("readWriteLock")
    private final FileSystem fileSystem;
```
nit: there's an annotation used for labeling the members that need to be guarded by a lock. I forgot to mention that earlier. Sorry for that

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,248 @@
+    public void createDirtyResult(WithJobResult jobResultEntry) throws IOException {
+        final Path path = constructDirtyPath(jobResultEntry.getJobId());
+
+        Preconditions.checkState(
```
According to the (updated) `JobResultStore` contract, we want to fail even if there is a clean entry available.
```
@throws IllegalStateException if the passed {@code jobResultEntry} has a {@code JobID}
     *     attached that is already registered in this {@code JobResultStore}.
```

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,248 @@
+
+    @Override
+    public void markResultAsClean(JobID jobId) throws IOException, NoSuchElementException {
+        Preconditions.checkState(
```
The `JobResultStore` contract changed here a bit as well. See the method's JavaDoc:
```
No actions should be triggered if the passed {@code JobID} belongs to a job that was already marked as clean.
```

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,248 @@
+     * @return The job result entry for the given Job ID.
+     */
+    @VisibleForTesting
+    protected JobResultEntry getJobResultEntry(JobID jobId)
```
This method is not used anywhere. We can remove it...

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,248 @@
+ * An implementation of the {@link JobResultStore} which persists job result data to an underlying
+ * distributed filesystem.
+ */
+public class FileSystemJobResultStore implements JobResultStore {
```
I noticed that `hasJobResultEntry` is not implemented. We might want to implement it instead of relying on the default implementation due to thread-safety. Checking whether a JobResult is dirty or clean should be done in an atomic fashion. I'm wondering whether we should remove the default implementation entirely from the interface to avoid running into this issue again. Up to you...

XComp replied:
side note: We could also move the default implementation into the abstract subclass, if we would go for the `AbstractThreadsafeJobResultStore` implementation I mentioned in the other comment above.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,248 @@
+    }
+
+    @VisibleForTesting
+    protected Set<JobResult> getCleanResults() throws IOException {
```
This is actually a test utility method that shouldn't be part of the implementation. You could move this method as a into the test. The directory to traverse is available there as well.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,248 @@
+            final Set<JobResult> dirtyResults = new HashSet<>();
+            final FileStatus fs = fileSystem.getFileStatus(this.basePath);
+            if (fs.isDir()) {
+                FileStatus[] statuses = fileSystem.listStatus(this.basePath);
+                for (FileStatus s : statuses) {
+                    if (!s.isDir()) {
+                        String path = s.getPath().getPath();
+                        if (path.contains(DIRTY_SUFFIX)) {
```
There is a shorter way of iterating over a directory in Java using the `FilenameFilter` and `File.listFiles`:
```
final File[] dirtyFiles =
                    new File(this.basePath.getPath())
                            .listFiles((dir, name) -> name.endsWith(DIRTY_SUFFIX + ".json"));
            for (File dirtyFile : dirtyFiles) {
                final JobResultEntry entry = mapper.readValue(dirtyFile, JobResultEntry.class);
                dirtyResults.add(entry.getJobResult());
            }
```

autophagy replied:
Do you know if this method (of relying on `java.io.File`) will work for distributed filesystems, like s3? The reason that I relied on iterating over file statuses is that I assumed `filesystem` is an abstraction over all those supported file systems.

XComp replied:
ha, good point - I guess you're right :-D

autophagy replied:
Actually, I think this has exposed a problem in my implementation - `getDirtyResults` gets the path and then uses the objectmapper to load the object by creating a `File`, which I dont think would work with distributed filesystems. I'll change that to instead use `FSDataInputStream`.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,248 @@
+            return fileSystem.exists(constructCleanPath(jobId));
+        } finally {
+            readWriteLock.readLock().unlock();
+        }
```
nit: just for the future - `hasCleanJobResultEntry` and `hasDirtyJobResultEntry` look quite alike. One could move the code into a separate method expecting a path as a parameter which then could be called in both methods. ...but that's always a per-case decision. Sometimes it's also not worth it to create an extra method if it reduces code-readability...

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,220 @@
+/** Tests for the {@link FileSystemJobResultStore}. */
+public class FileSystemJobResultStoreTest extends TestLogger {
+
+    @Rule public TemporaryFolder temporaryFolder = new TemporaryFolder();
```
It's done slightly differently in JUnit5.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,220 @@
+
+    private FileSystemJobResultStore fileSystemJobResultStore;
+
+    private static final JobResultEntry DUMMY_JOB_RESULT_ENTRY = createDummyJobResultEntry();
```
nit: Good habit is to move static final fields into the top of the declaration area.

XComp commented:
Thanks @autophagy for addressing my changes. I went over the PR and added some (mostly minor) comments. I had one idea about the custom serialization which I mentioned in one of the threads. See my comments above.

XComp committed:
[hotfix] Migrates ApplicationDispatcherBootstrapITCase to JUnit5/AssertJ

XComp committed:
[hotfix] Introduces TestingPartialDispatcherServices

autophagy commented:
@XComp Thanks for the review! I think i've addressed your comments. A few things:
- I played around with embedding the description of the config options in the `FilesystemJobResultStore` but it felt a little awkward, i'm not sure if this works so well. But I also take your point about the docs drifting if the implementation changes.
- I abstracted out the locking logic into an `AbstractThreadsafeJobResultStore` in https://github.com/XComp/flink/pull/5/commits/d189cce066d32ea09db720e5bf0c23dc60b4c5aa - I wanted to do it the same way the `EmbeddedJobStore` did with `withWriteLock` and `withReadLock` functions to cut down repetition, but the presence of `throws IOException` meant that I had to have a try/catch within the lambdas that are passed to `withRead/WriteLock` which removed any benefit from cutting down repetition. Do you have any advice here? The `throws IOException` everywhere feels a little messy, but im not sure if there's a different way of doing it.
- For the de/serialization I created a subclass of `JobResultEntry` called `JsonJobResultEntry` in https://github.com/XComp/flink/pull/5/commits/0a07cdfd33d28b7ac3f2065102ec8c7b8258aa16 - but i'm not sure if this is what you meant by having a copy constructor. It feels like it's doubling the implementation, which feels like a code smell. But again, any guidance you have here would be appreciated, I'm unused to these idioms :smile:

XComp committed:
[hotfix] Adds TestLogger extension to ApplicationDispatcherBootstrapTest

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,253 @@
+
+    @TempDir File temporaryFolder;
+
+    private static final ObjectMapper mapper = new ObjectMapper();
```
nit: private static final members can be moved to the top and should have a all-capitalized name

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,165 @@
+
+    private final boolean deleteOnCommit;
+
+    FileSystemJobResultStore(FileSystem fileSystem, Path basePath, boolean deleteOnCommit)
```
The `@VisibleForTesting` annotation is still valid here

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/JobResultEntry.java:
```diff
@@ -0,0 +1,81 @@
+     * Wrapper class around {@link JobResultEntry} to allow for serialization of a schema version,
+     * so that future schema changes can be handled in a backwards compatible manner.
+     */
+    public static class JsonJobResultEntry {
```
```suggestion
    public static class JsonJobResultEntry extends JobResultEntry {
```
Additionally, you can extract that class into its own file or even make it a private internal class of the `FileSystemJobResultStore`.

XComp replied:
FYI: What you did was creating an inner class (i.e. a class within a class), not a subclass (a class inheriting from a parent class through `extends`)

XComp commented on flink-core/src/main/java/org/apache/flink/configuration/JobResultStoreOptions.java:
```diff
@@ -0,0 +1,46 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.configuration;
```
What was the reason for you to move the `JobResultStoreOptions` into `flink-core`? Not saying it's wrong. I just would like to understand the reason...

XComp reviewed (commented):
Thanks @autophagy . We're close to finalizing this PR. Good job. :-) I added a few comments below. For a final run over the changes, it would be great to get the branch rebased after I force-pushed the FLINK-25430 base branch 😇

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedJobResultStore.java:
```diff
@@ -0,0 +1,93 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(EmbeddedJobResultStore.class);
+
+    @GuardedBy("readWriteLock")
```
The annotation should be moved into `AbstractThreadsafeJobResultStore` as well. We would need to annotate the `*Internal` methods instead, AFAIK

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedJobResultStore.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @Override
+    public void markResultAsCleanInternal(JobID jobId) throws IOException, NoSuchElementException {
```
Here, I'm wondering whether we could make the locking more finegrained. The initial Preconditions check resulting in the debug log message (which is guarded by a read lock) is something we could move into the abstract parent class as well. The actual `markResultAsCleanInternal` should be guarded by the write lock and only do the actual dictionary access. ...similar to how it's done for `createDirtyResult` already

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedJobResultStore.java:
```diff
@@ -0,0 +1,93 @@
+    final Map<JobID, WithJobResult> cleanJobResults = new HashMap<>();
+
+    @Override
+    public void createDirtyResultInternal(WithJobResult jobResultEntry) throws IOException {
```
```suggestion
```
Additionally, the `IOExceptions` in the different method signatures can be removed. The `EmbeddedJobResultStore` does not throw any `IOException`. This would also cleanup the diff in the `EmbeddedJobResultStoreTest`.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,253 @@
+        Path path = new Path(temporaryFolder.toURI());
+        fileSystemJobResultStore = new FileSystemJobResultStore(path.getFileSystem(), path, false);
+    }
+
```
What about adding a test that verifies that the base directory is created during initialization?

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,253 @@
+                        .build());
+    }
+
+    private List<JobID> getCleanResultIds() throws IOException {
```
```suggestion
    private List<JobID> getCleanResultIdsFromFileSystem() throws IOException {
```
nit: Being more explicit in the method's name about what the method does helps reading the code later one (reduces the necessity to actually look where the information is coming from).

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,253 @@
+     * into a clean state.
+     */
+    @Test
+    public void testCleanDirtyJobResult() throws Exception {
```
We might also want to verify here that the dirty file does not exist anymore.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,253 @@
+     * into a clean state.
+     */
+    @Test
+    public void testCleanDirtyJobResultTwice() throws Exception {
```
This test tests the same thing the previous test method `testCleanDirtyJobResult` tests. Additionally, the test method name indicates that we want to test something else here.

autophagy committed:
Refactor JRS Config Options

autophagy committed:
More finegrained locking, move common precondition check to abstract class

autophagy committed:
Make JsonJobResultEntry an internal class of FilesystemJobResultStore

autophagy committed:
Refactor tests

autophagy commented:
@XComp Thank you for the feedback and pairing! :smile: I've pushed up some commits to address the comments - for the testing I added assertions of expected file existence/nonexistence. It looks a little verbose, but I think its nice to be clear as to what state we expect to be in over the course of a test. I'll look more into the `TestingFileSystem` to see if it would give us any added benefit over a base filesystem.

XComp committed:
[FLINK-25430] Add JobResultStore

This commit adds the JobResultStore interface, a in-memory
implementation EmbeddedJobResultStoreand, and corresponding
test, container and utility classes.

XComp committed:
[FLINK-25430] Adds JobResultStore to HighAvailabilityServices and integrates it into the Dispatcher

XComp committed:
[hotfix] Migrates SessionDispatcherLeaderProcessTest to JUnit5/AssertJ

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/JobResultStoreOptions.java:
```diff
@@ -0,0 +1,58 @@
+                            "Determines whether job results should be automatically removed "
+                                    + "from the underlying job result store when they are clean. If "
+                                    + "false, the cleaned job results are, instead, marked as clean "
+                                    + "to indicate their state.");
```
```suggestion
                                    + "to indicate their state. Flink loses ownership in that case, i.e. the resources need to be cleaned up by the user.");
```
nit: Just something to describe the consequence to the user

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/AbstractThreadsafeJobResultStore.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    @GuardedBy("readWriteLock")
+    public abstract void createDirtyResultInternal(JobResultEntry jobResultEntry)
```
```suggestion
    protected abstract void createDirtyResultInternal(JobResultEntry jobResultEntry)
```
The `*internal` methods can be marked as `protected`. They are not part of the public `JobResultStore` interface and are only used internally by subclasses.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/AbstractThreadsafeJobResultStore.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    @GuardedBy("readWriteLock")
+    public abstract void markResultAsCleanInternal(JobID jobId)
```
```suggestion
    protected abstract void markResultAsCleanInternal(JobID jobId)
```

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/AbstractThreadsafeJobResultStore.java:
```diff
@@ -0,0 +1,130 @@
+    public boolean hasJobResultEntry(JobID jobId) throws IOException {
+        readWriteLock.readLock().lock();
+        try {
+            return hasDirtyJobResultEntry(jobId) || hasCleanJobResultEntry(jobId);
```
```suggestion
            return hasDirtyJobResultEntryInternal(jobId) || hasCleanJobResultEntryInternal(jobId);
```
nit: not really necessary because of the lock being reentrant. But using the `*Internal` methods would avoid going through the lock routine once more here. On our case it's not really a performance issue because the method won't be called that regularly. I just wanted to point it out...

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/AbstractThreadsafeJobResultStore.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    @GuardedBy("readWriteLock")
+    public abstract boolean hasDirtyJobResultEntryInternal(JobID jobId) throws IOException;
```
```suggestion
    protected abstract boolean hasDirtyJobResultEntryInternal(JobID jobId) throws IOException;
```

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/AbstractThreadsafeJobResultStore.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    @GuardedBy("readWriteLock")
+    public abstract boolean hasCleanJobResultEntryInternal(JobID jobId) throws IOException;
```
```suggestion
    protected abstract boolean hasCleanJobResultEntryInternal(JobID jobId) throws IOException;
```

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/AbstractThreadsafeJobResultStore.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    @GuardedBy("readWriteLock")
+    public abstract Set<JobResult> getDirtyResultsInternal() throws IOException;
```
```suggestion
    protected abstract Set<JobResult> getDirtyResultsInternal() throws IOException;
```

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedJobResultStore.java:
```diff
@@ -20,112 +20,57 @@
-                jobResultEntry.getJobId());
-
-        withWriteLock(() -> dirtyJobResults.put(jobResultEntry.getJobId(), jobResultEntry));
+    public void createDirtyResultInternal(JobResultEntry jobResultEntry) {
```
```suggestion
    protected void createDirtyResultInternal(JobResultEntry jobResultEntry) {
```

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedJobResultStore.java:
```diff
@@ -20,112 +20,57 @@
-            LOG.debug("The job {} is already marked as clean. No action required.", jobId);
-
-            return;
+    public void markResultAsCleanInternal(JobID jobId) throws IOException, NoSuchElementException {
```
```suggestion
    protected void markResultAsCleanInternal(JobID jobId) throws IOException, NoSuchElementException {
```

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedJobResultStore.java:
```diff
@@ -20,112 +20,57 @@
     @Override
-    public boolean hasDirtyJobResultEntry(JobID jobId) {
-        return withReadLock(() -> dirtyJobResults.containsKey(jobId));
+    public boolean hasDirtyJobResultEntryInternal(JobID jobId) throws IOException {
```
```suggestion
    protected boolean hasDirtyJobResultEntryInternal(JobID jobId) throws IOException {
```

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedJobResultStore.java:
```diff
@@ -20,112 +20,57 @@
     @Override
-    public boolean hasCleanJobResultEntry(JobID jobId) {
-        return withReadLock(() -> cleanJobResults.containsKey(jobId));
+    public boolean hasCleanJobResultEntryInternal(JobID jobId) throws IOException {
```
```suggestion
    protected boolean hasCleanJobResultEntryInternal(JobID jobId) throws IOException {
```

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedJobResultStore.java:
```diff
@@ -20,112 +20,57 @@
-        } finally {
-            readWriteLock.readLock().unlock();
-        }
+    public Set<JobResult> getDirtyResultsInternal() throws IOException {
```
```suggestion
    protected Set<JobResult> getDirtyResultsInternal() throws IOException {
```

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,210 @@
+        if (isNullOrWhitespaceOnly(jrsStoragePath)) {
+            final String haStoragePath = config.get(HighAvailabilityOptions.HA_STORAGE_PATH);
+            final String haClusterId = config.get(HighAvailabilityOptions.HA_CLUSTER_ID);
+            basePath = new Path(haStoragePath + "/job-result-store/" + haClusterId);
```
I know that you found a generic solution for it in the documentation already. That's sufficient enough. I thought about it once more and just want to lay out a different approach (which you don't have to follow; just for future tasks). 

If you would have created a static method `FileSystemJobResultStore.createDefaultPath(String baseDirPath, String clusterId)` like:
```
public static String createDefaultJobResultStorePath(String baseDir, String clusterId) {
    return baseDir + "/job-result-store/" + clusterId;
}
```
which you could have used here
```suggestion
            basePath = new Path(createDefaultJobResultStorePath(haStoragePath, haClusterId));
```
you could have also used `FileSystemJobResultStore.createDefaultJobResultStore(String.format("{%s}", HighAvailabilityOptions.HA_STORAGE_PATH.key()), String.format("{%s}", HighAvailabilityOptions.HA_CLUSTER_ID.key())` in the description of `JobResultStoreOptions.STORAGE_PATH` to specify the default path without having the path multiple times in the codebase.

autophagy replied:
This is a really good idea, I will implement this!

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,210 @@
+     * @return A path for a clean entry for the given the Job ID.
+     */
+    private Path constructCleanPath(JobID jobId) {
+        return new Path(this.basePath.getPath() + "/" + jobId.toString() + ".json");
```
```suggestion
        return new Path(this.basePath.getPath(), jobId.toString() + ".json");
```
nit: this omits the file separator which is used by `Path`.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,210 @@
+     * @return A path for a dirty entry for the given the Job ID.
+     */
+    private Path constructDirtyPath(JobID jobId) {
+        return new Path(this.basePath.getPath() + "/" + jobId.toString() + DIRTY_SUFFIX + ".json");
```
```suggestion
        return new Path(this.basePath.getPath(), jobId.toString() + DIRTY_SUFFIX + ".json");
```
nit

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,210 @@
+    @Override
+    public void markResultAsCleanInternal(JobID jobId) throws IOException, NoSuchElementException {
+        Path dirtyPath = constructDirtyPath(jobId);
+        if (fileSystem.exists(dirtyPath)) {
+            if (deleteOnCommit) {
+                fileSystem.delete(dirtyPath, false);
+            } else {
+                fileSystem.rename(dirtyPath, constructCleanPath(jobId));
+            }
+        } else {
+            throw new NoSuchElementException(
+                    String.format(
+                            "Could not mark job %s as clean as it is not present in the job result store.",
+                            jobId));
+        }
```
```suggestion
        if (!fileSystem.exists(dirtyPath)) {
            throw new NoSuchElementException(
                    String.format(
                            "Could not mark job %s as clean as it is not present in the job result store.",
                            jobId));
        }
        
        if (deleteOnCommit) {
            fileSystem.delete(dirtyPath, false);
        } else {
            fileSystem.rename(dirtyPath, constructCleanPath(jobId));
        }
```
nit: You could remove the nested if by inverting the if condition which might improve readability. But again, like most of the other comments - it's more like a cosmetic change and up to your taste.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,210 @@
+            FileStatus[] statuses = fileSystem.listStatus(this.basePath);
+            for (FileStatus s : statuses) {
+                if (!s.isDir()) {
+                    String path = s.getPath().getPath();
+                    if (path.contains(DIRTY_SUFFIX)) {
```
```suggestion
                    final String filename = s.getPath().getName();
                    if (filename.contains(DIRTY_SUFFIX)) {
```
nit: I don't know - this `getPath().getPath()` looks weird. What about using the filename instead? That might be sufficient enough.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,285 @@
+        assertThat(expectedDirtyFile(DUMMY_JOB_RESULT_ENTRY)).doesNotExist();
+        assertThat(expectedCleanFile(DUMMY_JOB_RESULT_ENTRY)).doesNotExist();
+    }
+
```
```suggestion
    @Test
    public void testVersionSerialization() throws IOException {
        fileSystemJobResultStore.createDirtyResult(DUMMY_JOB_RESULT_ENTRY);
        final File dirtyFile = expectedDirtyFile(DUMMY_JOB_RESULT_ENTRY);
        final JsonJobResultEntry deserializedEntry =
                MAPPER.readValue(dirtyFile, JsonJobResultEntry.class);

        assertThat(dirtyFile).isFile().content().containsPattern("\"version\":1");
    }

    @Test
    public void testJobResultSerializationDeserialization() throws IOException {
        fileSystemJobResultStore.createDirtyResult(DUMMY_JOB_RESULT_ENTRY);
        final File dirtyFile = expectedDirtyFile(DUMMY_JOB_RESULT_ENTRY);
        final JsonJobResultEntry deserializedEntry =
                MAPPER.readValue(dirtyFile, JsonJobResultEntry.class);

        final JobResult deserializedJobResult = deserializedEntry.getJobResult();
        assertThat(deserializedJobResult)
                .extracting(JobResult::getJobId)
                .isEqualTo(DUMMY_JOB_RESULT_ENTRY.getJobId());
        assertThat(deserializedJobResult)
                .extracting(JobResult::getApplicationStatus)
                .isEqualTo(DUMMY_JOB_RESULT_ENTRY.getJobResult().getApplicationStatus());
        assertThat(deserializedJobResult)
                .extracting(JobResult::getNetRuntime)
                .isEqualTo(DUMMY_JOB_RESULT_ENTRY.getJobResult().getNetRuntime());
        assertThat(deserializedJobResult)
                .extracting(JobResult::getSerializedThrowable)
                .isEqualTo(DUMMY_JOB_RESULT_ENTRY.getJobResult().getSerializedThrowable());
        assertThat(deserializedJobResult)
                .extracting(JobResult::getAccumulatorResults)
                .isEqualTo(DUMMY_JOB_RESULT_ENTRY.getJobResult().getAccumulatorResults());
    }
    
```
We might want to add another test to verify that the version is serialized and the JobResult is serialized properly.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,210 @@
+        private static final String FIELD_NAME_RESULT = "result";
+        private static final String FIELD_NAME_VERSION = "version";
+
+        JsonJobResultEntry(JobResultEntry entry) {
```
```suggestion
        private JsonJobResultEntry(JobResultEntry entry) {
```

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java:
```diff
@@ -0,0 +1,210 @@
+        }
+
+        @JsonCreator
+        public JsonJobResultEntry(@JsonProperty(FIELD_NAME_RESULT) JobResult jobResult) {
```
```suggestion
        private JsonJobResultEntry(@JsonProperty(FIELD_NAME_RESULT) JobResult jobResult) {
```

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,285 @@
+    /** Tests that the base directory is created on Filesystem Job Result Store initialization. */
+    @Test
+    public void testBaseDirectoryCreationOnResultStoreInitialization() throws Exception {
+        final File tempDir = new File(temporaryFolder.getPath() + "/new-temp-dir");
```
```suggestion
        final File tempDir = new File(temporaryFolder.getPath(), "new-temp-dir");
```
nit: to avoid this separator specification

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,285 @@
+    /** Tests that the base directory is created on Filesystem Job Result Store initialization. */
+    @Test
+    public void testBaseDirectoryCreationOnResultStoreInitialization() throws Exception {
+        final File tempDir = new File(temporaryFolder.getPath() + "/new-temp-dir");
+        final Path tempPath = new Path(tempDir.getPath());
```
```suggestion
        final File baseDir = new File(temporaryFolder.getPath() + "/new-temp-dir");
        final Path baseDirPath = new Path(tempDir.getPath());
```
"tempDir" is a bit overloaded in this test class due to the `temporaryFolder`. Let's be more specific about the purpose of this folder.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,285 @@
+        fileSystemJobResultStore = new FileSystemJobResultStore(path.getFileSystem(), path, false);
+    }
+
+    /** Tests that the base directory is created on Filesystem Job Result Store initialization. */
```
The JavaDoc is obsolete here. Usually, we describe the test case through the test method name which is also good enough in that case. That also applies to the other test methods below...

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,285 @@
+    @Test
+    public void testStoreDirtyJobResultTwice() throws IOException {
+        fileSystemJobResultStore.createDirtyResult(DUMMY_JOB_RESULT_ENTRY);
+        assertThat(expectedDirtyFile(DUMMY_JOB_RESULT_ENTRY)).exists().isFile().isNotEmpty();
```
```suggestion
```
nit: That is already tested in `testStoreDirtyJobResult` and is out-of-scope for this test case. We're interested in the second `createDirtyResult` call here. Same goes for the file creation after `markAsClean` is called.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,285 @@
+    }
+
+    private static JobResultEntry createDummyJobResultEntry() {
+        return new JobResultEntry(
```
`TestingJobResultStore` provides `DUMMY_JOB_RESULT` now in addition to `JobResult` factory methods.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,285 @@
+     */
+    private File expectedDirtyFile(JobResultEntry entry) {
+        return new File(
+                temporaryFolder.toURI().getPath()
+                        + "/"
+                        + entry.getJobId().toString()
+                        + "_DIRTY.json");
```
```suggestion
        return new File(temporaryFolder, entry.getJobId().toString() + "_DIRTY.json");
```
nit: can be shortened

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStoreTest.java:
```diff
@@ -0,0 +1,285 @@
+     * @return The expected clean file.
+     */
+    private File expectedCleanFile(JobResultEntry entry) {
+        return new File(
+                temporaryFolder.toURI().getPath() + "/" + entry.getJobId().toString() + ".json");
```
```suggestion
        return new File(temporaryFolder, entry.getJobId().toString() + ".json");
```
nit: shorter version

XComp reviewed (commented):
Good job. :-) 👍 I was a bit nitty in the final round. But most of the comments are really on of cosmetic nature. I'm still a bit torn whether we should combine the two `*JobResultStoreTest` implementations. They have quite a bit of redundant code. But I'm also not a big fan of inheritance in test code. Because it makes reading the test harder. So, I left it like that.

Looks like we're ready to merge after the few comments are addressed, the commits are squashed and the branch is rebased after [FLINK-25430 PR](https://github.com/apache/flink/pull/18189) is approved. Great! 🎉

autophagy commented:
@XComp Thank you a ton for all your guidance and wisdom on this! Feel like I've learned a lot in the process :slightly_smiling_face:. I've addressed all your comments, but will instead close this PR and open a new one against apache/flink with the results squashed into a logical commit.
